### ISSUE TITLE
Alternative Implementation of Named Parameters

### ISSUE THREAD

**Issue by eviltwinskippy** [flightaware/Tcl-bounties#30]
I have started work on a new TIP: <https://core.tcl.tk/tips/doc/trunk/tip/479.md>

This tip is based on using a new command **@proc** to prototype a dictionary based descriptor for named parameters. With some testing, it will eventually add a new facility to **proc** when the developer specifies args.

	% @proc p {
             a {default {}}
             b {default {}}
         } {
           return [list a: $a b: $b $args] 
        }
	% p
	> a: {} b: {} {}
	% p foo bar
	> a: {} b: {} {foo bar}

Or

	% proc q {
             {a {}}
             {args {
               b {default {}}
             }
          } { 
            return [list a: $a b: $b $args] 
          }
	% q
	> a: {} b: {} 
	% q foo bar baz
	> a: foo b: {} {bar baz}

This tip is under development as a pure-tcl extension as well as a TEA based extension at: <http://fossil.etoyoc.com/fossil/tip479/doc/trunk/index.md>

When ready for integration into the core, this tip will introduced as the tip479 branch to trunk.

**Comment by resuna** [comment]
This seems to be incompatible with existing proc.

**Comment by eviltwinskippy** [comment]
The only modification to existing argument handling for procs is that the "args" placeholder at the end can take an argument containing the specs for named arguments.

proc myproc args {} 

Works the same as ever, unmodified

proc myproc {{args {a {} b {}}}}

Triggers the new named parameters behavior, where a and b are treated as named parameters that are expected on every call. Naturally this would be a bit cumbersome for the general use of named parameters, so the standard provides a shortcut:

@proc mproc {a {} b {}} 

Which cuts to the chase.

**Comment by resuna** [comment]
So in the call, the optional arguments can only come at the end?

**Comment by eviltwinskippy** [comment]
In calls to proc, named arguments can only come at the end. And they are only named parameters if the args keyword is given an argspec. (Otherwise extra arguments are treated as a list.)

The @procs command takes named parameters only. Whether a parameter is optional or not depends on how each parameter was configured. 
```
@proc myproc {
  a {mandatory 1}
  b {default {nob}}
  c {mandatory 0}
} {
  # Because it is not mandatory, $c may not exist
  if {[dict exists $args c]} {
    puts [list $a $b $c]
  } else {
    puts [list $a $b]
}
```


```
proc myproc {args {
  a {mandatory 1}
  b {default {nob}}
  c {mandatory 0}
}} {
  # Because it is not mandatory, $c may not exist
  if {[dict exists $args c]} {
    puts [list $a $b $c]
  } else {
    puts [list $a $b]
}
```

**Comment by eviltwinskippy** [comment]
The nice thing is that positional arguments can be added ex-post-facto to existing commands via the @args command.

```
proc oldproc {sender subject timestamp args} {
  @args {reply-to {default noreply@mydomain.com}} $args
}
```

**Comment by eviltwinskippy** [comment]
Also useful for ensembles:

```
proc myensemble {command args} {
  switch $command {
    smtp {
       # Dicts can be created on the fly
       @args [dict create from {} to {} mtime [dict create default [clock seconds]]] $args
       smtp::dispatch $from $do mtime $mtime
     }
     http {
       # Or accept a static string with whitespace
       @args {
          from {}
          to {}
          url {default http://mydomain.info}
       } $args
       http::post ${url}?from=${from}&to=${to}
     }
  }
```

**Comment by eviltwinskippy** [comment]
Also, what isn't reserved is fair game:
```

@proc prettyproc {
  color {
    comment {Color of the button}
    type color
    default green
  }
  flavor {
    comment {What flavor candy emerges when pressed.}
    default random
    options {vanilla caramel chocolate strawberry}
  }
} {
   set flavors  [dict get ${@spec} flavor options]
   if {$flavor eq "random} { 
     set flavor [lrandom $flavors]
   } else {
     if {$flavor ni $flavors} {
        error "Invalid flavor $flavor. Valid: random OR $flavors"
     }
  }
}
```

**Comment by eviltwinskippy** [comment]
Updates: The branch promised for Tip#479 has been added to the core, and based on some community feedback, the concept has been altered.

The @proc keyword has been replaced with ::dictargs::proc
The @args keyword has been replaced with ::dictargs::parse

Instead of simply storing the argument spec as the default of args, the default of args now has a first argument telling the interp what is the namespace with the rules for the specs, followed by the actual argument spec.

This last modification is to allow for multiple rule sets to exist. For instance a library of tools may use both dictargs and something like Ruby Lane's parse_args.

For evaluation purposes, I have prepared a TEA extension which actually gives 95% of the performance of the core modification inside of an unmodified 8.6 core. This implementation is also available as a plug-in for the Practcl kit wrapper.
